db/persistent_stats_history.cc

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+#include "db/persistent_stats_history.h"
+
+#include <string>
+#include <utility>
+#include "db/db_impl/db_impl.h"
+#include "port/likely.h"
+#include "util/string_util.h"
+
+namespace rocksdb {
+// 10 digit seconds timestamp => [Sep 9, 2001 ~ Nov 20, 2286]
+const int kNowSecondsStringLength = 10;
+const int kMicrosInSecond = 1000 * 1000;
+const char* kVersionKeyString = "__persistent_stats_version__";
+const int kPersistentStatsVersion = 1;
+
+PersistentStatsHistoryIterator::~PersistentStatsHistoryIterator() {}
+
+bool PersistentStatsHistoryIterator::Valid() const { return valid_; }
+
+Status PersistentStatsHistoryIterator::status() const { return status_; }
+
+void PersistentStatsHistoryIterator::Next() {
+  // increment start_time by 1 to avoid infinite loop
+  AdvanceIteratorByTime(GetStatsTime() + 1, end_time_);
+}
+
+uint64_t PersistentStatsHistoryIterator::GetStatsTime() const { return time_; }
+
+const std::map<std::string, uint64_t>&
+PersistentStatsHistoryIterator::GetStatsMap() const {
+  return stats_map_;
+}
+
+int PersistentStatsHistoryIterator::EncodeVersionKey(uint64_t now_micros,
+                                                     int size, char* buf) {
+  // make time stamp string equal in length to allow sorting by time
+  int length = snprintf(buf, size, "%010d#%s",
+                        static_cast<int>(now_micros / kMicrosInSecond),
+                        kVersionKeyString);
+  return length;
+}
+
+int PersistentStatsHistoryIterator::DecodeVersionNumber(
+    const char* timestamp_str, Iterator* iter) {
+  char version_key[100];
+  snprintf(version_key, 100, "%s#%s", timestamp_str, kVersionKeyString);
+  iter->Seek(version_key);
+  // cannot find version_number, could be a user-defined CF or internal error
+  // skip all with the same timestamp
+  if (!iter->Valid()) {
+    return -1;
+  } else {
+    // read version_number but do nothing in current version
+    int version_number =
+        static_cast<int>(std::stoull(iter->value().ToString()));
+    // reset iter to beginning
+    iter->Seek(timestamp_str);
+    return version_number;
+  }
+}
+
+int PersistentStatsHistoryIterator::EncodeKey(uint64_t now_micros,
+                                              const std::string& key, int size,
+                                              char* buf) {
```

Comment:
I don't get why encoder belongs to iterator. Iterator is only for read, right?

Reply:
Yes but you said earlier 
> I prefer the key encoding code is placed closer to key decoding code. It makes easier for people to read the code and reason about correctness.

will think about a better place to put them